### ISSUE TITLE
Ability to control token of KendoTemplateBehavior of Kendo ComboBox

### ISSUE THREAD

**Issue by reckart** [sebfz1/wicket-jquery-ui#245]
The KendoTemplateBehavior of the Kendo ComboBox uses a random number (actually the hash of the KendoTemplateBehavior) as a token for the Javascript header item. If a ComboBox is often replaced via AJAX on a page, this causes Javascript header items for KendoTemplateBehavior to accumulate in the page.

On the one hand, this is good, because each new instance of a ComboBox may have a different template. 

However, on the other hand this is bad, because of the accumulation. 

It would be nice if KendoTemplateBehaviors of ComboBoxes that have been removed from a page (even if in a repeater) could be cleaned up automatically.

It would also be nice if one could control/influence the ID of the KendoTemplateBehaviors. E.g. the KendoDataSource JS Token is derived from the markup ID of the host component. I am currently using this fact in to clean up stale JS datasource declarations.

**Comment by sebfz1** [comment]
Hi Richard,

`KendoTemplateBehavior` now implements `IDestroyable` and will run the following statement on ajax component reload:

```javascript
(function(){var $w = jQuery('#combobox26-template'); if($w) { $w.detach(); }})();
```

Please also note that the template html-id is now based on the hosting component markup-id.
Hope this will suit your needs.

I will deploy a snapshot so you can test before the next incoming release...

Best regards,
Sebastien.

**Comment by reckart** [comment]
@sebfz1 sounds good. I didn't know about IDestroyable :) 

If the component cleans up after itself, I might not even have to do that manually (i.e. in that case, I might not need to know/be able to construct the ID). But it is probably still a good idea being able to control it.

Can the automatic cleaning up also be done for the KendoDataSource of the ComboBoxBehavior?

**Comment by sebfz1** [comment]
> Can the automatic cleaning up also be done for the KendoDataSource of the ComboBoxBehavior?

Unfortunately `KendoDataSource` is not a `Behavior` so it cannot implement `IDestroyable`. I am currently changing naming strategy. I will see if/how it can be destroyed too...

> in that case, I might not need to know/be able to construct the ID

That's that I guessed. But I'm finding this cleaner now :)

**Comment by sebfz1** [comment]
I've reworked the `KendoDataSource` naming and added 2 ctors (taking a `Component`)

```java
public KendoDataSource(Component component) { }
public KendoDataSource(Component component, String type) { }
```

The datasource name, if issued from a component, is of the form:
`component.getMarkupId() + "_datasource";`

Which leads to:

```javascript
<script id="kendo-combobox11_datasource" type="text/javascript">
jQuery(function() { window.combobox11_datasource = new kendo.data.DataSource({ /* ... */ }); });
</script>
```

Note that the script-id (token) is still prefixed by `kendo-` to prevent confusion with the datasource variable name (even not in same scope)

I've also added `KendoDataSource#destroy` method which is automatically called. The statement look like:

```javascript
(function(){ var $w = jQuery('#kendo-combobox11_datasource'); if($w) { $w.detach(); }})();
```

**Comment by sebfz1** [comment]
Hi Richard,

Forgot to mention I've deployed `7.4.1-SNAPSHOT`

Best regards,
Sebastien

**Comment by reckart** [comment]
I have downgraded again from 7.4.0 to 6.24.0 because some users preferred not being forced to upgrade to Tomcat 7. But I can temporarily upgrade back to 7.4.1-SNAPSHOT for testing. If all works, are you planning on a 6.24.1?

**Comment by sebfz1** [comment]
Code is aligned on both 6.x & 7.x branches.
6.25.0 will soon be released but I will deploy a snapshot so you can give it a try...

**Comment by sebfz1** [comment]
Deployed `6.24.1-SNAPSHOT`

Kendo UI has been upgraded to `2016.3.914.core`
There is some API breaks:
```
**API BREAK** Kendo UI: Updated AbstractWindow#setTitle(IModel<String> title) to #setTitle(String title) 
**API BREAK** Kendo UI: AbstractWindow#getTitle now returns a String instead of an IModel<String> 
**API BREAK** Kendo UI: Removed TreeViewModelBehavior#newRequestHandler(), should override #getResponse instead
**API BREAK** Kendo UI: Moved repeater/dataview/DataProviderBehavior to repeater/
**API BREAK** Kendo UI: ListViewBehavior, added KendoDataSource in ctor signature
```

**Comment by martin-g** [comment]
Hi Sebastien,

What was the motivation for this change ?
> **API BREAK** Kendo UI: Updated AbstractWindow#setTitle(IModel<String> title) to #setTitle(String title)

**Comment by sebfz1** [comment]
I've added setTitleModel and getTitleModel...

**Comment by sebfz1** [comment]
Doh, I didn't saw it was you Martin! :)

**Comment by reckart** [comment]
Hm... I have switched to 6.24.1-SNAPSHOT. However, it does not quite work as expected.

The first time I load the page and perform an action that causes the a KendoComboBox to be rendered, I see the datasource and template elements being added to the head. When I perform any action that triggers an AJAX reload of the same combobox or when the comboboxes are replaced by others, then I see only short flicker in the head (probably the datasource/template appearing and then immediately disappearing again).

It seems a bit as if something in the ordering of things is wrong... as if the cleanup action is performed on every AJAX request and potentially *after* a combobox has been replaced by a new one (i.e. removing the template/datasource of the *new* combobox instead of the old one).

Note that I'm overriding the markup ID, so e.g. if a combobox is replaced, then it can happen that the old combobox and the new one have the same markup ID but user different choices (i.e. data sources).

**Comment by sebfz1** [comment]
The cleanup is performed (or at least supposed to be performed) when a combobox is reloaded/reattached on an ajax request. The order should be good, you can verify in the ajax debug window (in the priority-evaluate part).

That should be be working case.
On another part, and as each widgets are self-destroyable, I don't know if it works if you replace one by another...

I will have a look, thanks!

**Comment by reckart** [comment]
I put the AJAX debug log into this [pastebin](http://pastebin.com/Nfxy3hAP).

Search e.g. for "featureEditorHead1420_template". It seems to me that the template is added to the header first and *then* the code that includes the "detach()" is sent.

**Comment by sebfz1** [comment]
Oh, I see. #destroy/#detach works as expected to element in the body. In case of it's in the header, then yes, we seems to have a problem because the header is added first in any case... I know there is some options to control header priority but I don't know if these can help here...
Thanks for the input, I will have a look tomorrow!

**Comment by sebfz1** [comment]
Hi Richard,

I believe there is no other choice than removing the template *before* the ajax call.
So, I've added the `#getTemplateToken()` method to component having a template, and an helping class `DetachAjaxCallListener`.

I don't see now a proper way to do it automatically, so the detach call remains on your side.

The usage if the following:

```java
form.add(new AjaxButton("button") {

	@Override
	protected void updateAjaxAttributes(AjaxRequestAttributes attributes)
	{
		attributes.getAjaxCallListeners().add(new DetachAjaxCallListener(combobox.getTemplateToken()));
	}
}
```

If you have several comboboxes, you can do this way too:

```java
form.add(new AjaxButton("button") {

	@Override
	protected void updateAjaxAttributes(AjaxRequestAttributes attributes)
	{
		AjaxCallListener listener = new AjaxCallListener();
		listener.onBeforeSend(JQueryUtils.getDetachStatement(combobox1.getTemplateToken()));
		listener.onBeforeSend(JQueryUtils.getDetachStatement(combobox2.getTemplateToken()));
		listener.onBeforeSend(JQueryUtils.getDetachStatement(comboboxX.getTemplateToken()));
	
		attributes.getAjaxCallListeners().add(listener);
	}
}
```

I am looking at datasources, and will inform you then the snapshot will be deployed...

Best regards,
Sebastien

**Comment by sebfz1** [comment]
Note to me: it is legal to have `script` elements in the `body`... (not sure I can use this, though)

**Comment by reckart** [comment]
Hm. The last suggestion adds a tight coupling between the button and the ComboBoxes. I would tend to prefer that they clean up themselves.

**Comment by sebfz1** [comment]
Hi Richard,

I would have used the event-bus, so it is catched by the comboxes themselves.
I did not have tested, but something like this should work:

```java
// to send the event:
BroadcastUtils.breadth(MyPage.this, new AjaxPayload(target));
```

```java
// to handle the event (MyComboBox):
@Override
public void onEvent(IEvent<?> event)
{
	super.onEvent(event);

	if (event.getPayload() instanceof AjaxPayload)
	{
		AjaxPayload payload = (AjaxPayload) event.getPayload();
		AjaxRequestTarget target = payload.getTarget();

		target.addListener(new IListener() {
			
			@Override
			public void onBeforeRespond(Map<String, Component> map, AjaxRequestTarget target)
			{
				target.prependJavaScript(JQueryUtils.getDetachStatement(MyComboBox.this.getTemplateToken()));
			}
		});
	}
}
```

`BroadcastUtils` has been be added recently IIRC, you might use the snapshot version or replace with a similar code:

```java
this.send(MyPage.this, Broadcast.BREADTH, new AjaxPayload(target));
```
